src/Components/test/testassets/TestServer/Startup.cs

```diff
@@ -38,24 +38,36 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
             }
 
             AllowCorsForAnyLocalhostPort(app);
-            app.UseMvc();
+
+            app.UseRouting();
 
             // Mount the server-side Blazor app on /subdir
             app.Map("/subdir", subdirApp =>
             {
                 // The following two lines are equivalent to:
-                //     subdirApp.UseServerSideBlazor<BasicTestApp.Startup>();
-                // However it's expressed using UseSignalR+UseBlazor as a way of checking that
-                // we're not relying on any extra magic inside UseServerSideBlazor, since it's
+                //     endpoints.MapComponentsHub<Index>();
+                //
+                // However it's expressed using routing as a way of checking that
+                // we're not relying on any extra magic inside MapComponentsHub, since it's
                 // important that people can set up these bits of middleware manually (e.g., to
                 // swap in UseAzureSignalR instead of UseSignalR).
-                subdirApp.UseRouting(routes =>
-                    routes.MapHub<ComponentHub>(ComponentHub.DefaultPath).AddComponent<Index>(selector: "root"));
+
+                subdirApp.UseRouting();
+
+                subdirApp.UseEndpoints(endpoints =>
+                {
+                    endpoints.MapHub<ComponentHub>(ComponentHub.DefaultPath).AddComponent<Index>(selector: "root");
```

Comment:
What are the rules for when there are multiple UseEndpoints? Is the scope that it is used in tracked, e.g. inside a Map, and the UseRouting for that scope gets routes from the correct UseEndpoints.

Reply:
Answered as I read through the rest of the PR 😄

Reply:
Rule 1: Don't do it, it's kinda nonsense
Rule 2: Why would you do this?
Rule 3: We do our best make it do something reasonable 😁